### PR TITLE
Fix legacy widget block preview iframe

### LINKED ISSUE #32281: Bug: Regression in core Legacy widget previews
Hey all, I noticed a regression in the way core Legacy widgets display as a preview in both widget editors, not sure when this may have merged. Instead of displaying a "no preview available" placeholder, they are displaying as a wp-admin top bar. 

Here's how they look on the Appearance Widgets screen:

Screen Shot 2021-05-27 at 10 26 05 AM

And on the Customizer Widgets screen:

Screen Shot 2021-05-27 at 11 25 37 AM

### PR DESCRIPTION
**Pull request by noisysocks**
### Description 

Fixes #32281.

The PHP code in `legacy-widget/index.php` registers the Legacy Widget block and intercepts any request with the `legacy-widget-preview` query param. This is how the block's preview iframe is implemented.

The iframe intercept code will not work if the Legacy Widget block has not been registered. This is because we render the iframe using `$block->render()`.

Block registration must be done in an `init` action beginning with `register_`. This is so that the Gutenberg plugin's webpack config can set the priority of this action to be 20 instead of 10. This lets the plugin redeclare blocks that are registered in core.

https://github.com/WordPress/gutenberg/blob/e4a2ea51c9bd5aa18e29918a963a1f80fe0be7d7/webpack.config.js#L275-L286

So, we have two requirements:

- The iframe intercept must happen after block registration; and
- Block registration must happen in an `init` action beginning with `register_`.

We could meet these two requirements by performing the iframe intercept in the `register_block_core_legacy_widget` function, but I don't like this because it means one cannot unhook registration without also unhooking the iframe intercept. It also means the function name doesn't describe what it does 😇

Instead, this PR performs the intercept in a second `init` action which uses the same priority as the first `init` action. This priority will be 20 in the plugin and 10 in core. We can determine an action's priority by using `has_action`.

https://developer.wordpress.org/reference/functions/has_action

### How to test

1. Go to Appearance → Widgets with the Gutenberg plugin installed and running latest WordPress `trunk`.
2. Insert or select a Legacy Widget block.
3. The preview functionality should work without displaying a WP Admin menu bar in the iframe.

**Note:** There is a PHP warning caused by registering `core/legacy-widget` twice. It's because core is incorrectly adding the registration action with priority 20. This will be fixed once `@wordpress/block-library` is published to npm and the changes in this PR are included in core. You can verify this by copying `packages/block-library/src/legacy-widget/index.php` in Gutenberg to `blocks/legacy-widget.php` in core.

### CODE REVIEW

**Comment by talldan** [comment]
> Go to Appearance → Widgets with the Gutenberg plugin installed and running latest WordPress trunk.

For me, the bug doesn't happen when I'm running the latest WordPress trunk. Any other steps to reproduce?

Happens every time in WordPress 5.7.3 with the plugin though. And I can confirm this fixes it.

> Note: There is a PHP warning caused by registering core/legacy-widget twice. It's because core is incorrectly adding the registration action with priority 20. This will be fixed once @wordpress/block-library is published to npm and the changes in this PR are included in core. You can verify this by copying packages/block-library/src/legacy-widget/index.php in Gutenberg to blocks/legacy-widget.php in core.

<strike>In WordPress trunk I get a white screen of death unfortunately. I'm unable to access any admin screen. I'm guessing that's not what you meant by a warning:</strike>
Screenshot 2021-05-28 at 4 33 22 pm

Actually I restarted the build and it seems to work now. Weird. Previews work ok (but they worked fine anyway for me), they just have the warning. Do you think this should be merged with the warning? It's quite disruptive. Is it possible to update core's version of `block-library` before releasing the plugin and shortly after merging this?

**Comment by noisysocks** [comment]
> > Go to Appearance → Widgets with the Gutenberg plugin installed and running latest WordPress trunk.
> 
> For me, the bug doesn't happen when I'm running the latest WordPress trunk. Any other steps to reproduce?

Huh, I could have swore it was happening for me on `trunk` on Friday. It works for me now too. Still, there's problem in that it looks like, with latest Core trunk + Gutenberg trunk, it uses Core's copy of `legacy-widgets/index.php`. With this branch applied, it correctly uses Gutenberg's copy of `legacy-widgets/index.php`. You can verify this by adding some debugging markup to the HTML at the bottom of `legacy-widgets/index.php` in Gutenberg. Plus, as you say, this bug exists when using 5.7.

> Do you think this should be merged with the warning? It's quite disruptive. Is it possible to update core's version of `block-library` before releasing the plugin and shortly after merging this?

Yeah, it's pretty gross. I can merge this PR and then publish an update to `@wordpress/block-library` immediately so that there's minimal disruption for Gutenberg contributors. Contributors would see the warning until their core is updated. We could patch core to remove the warning but this doesn't help as contributors would still see the warning until core is updated.

**Review by kevin940726 — APPROVED** [review]
I'm also seeing the white screen of death with the warning. I'm able to access the admin in the test(prod) environment though.

**Comment by noisysocks** [comment]
Thanks @kevin940726. Let's hold off on merging this until https://github.com/WordPress/wordpress-develop/pull/1319 is committed. Then, I'll merge this PR and immediately publish and update `@wordpress/block-library` in core.

**Comment by noisysocks** [comment]
I thought of a better way to fix this 😅

- `register_block_core_legacy_widget` should have priority 10 in Core and priority 20 in the plugin. This matches all other blocks. Removing the priority in `packages/block-library/src/legacy-widget/index.php` does exactly this.
- `handle_legacy_widget_preview_iframe` should always have priority 21. This ensures the block is registered before handling the iframe request.
- Add a temporary fix to the plugin which corrects the priority of Core's `register_block_core_legacy_widget` from 20 to 10. This fixes the "block is already registered" PHP warning.

I've updated this PR with those changes. Because of the temporary fix, we don't have to hold off on merging anymore. Instead, I will remove the temporary fix from the plugin in a week or so's time.

To test this, I browsed to Appearance → Widgets and verified that the Legacy Widget block loads and that its preview functionality works. I did this for multiple versions of Core with and without the plugin enabled.

| Core branch | Gutenberg disabled | Gutenberg enabled |
| - | - | - |
| `5.7` | N/A (the old screen loads) | ✅ |
| `trunk` | ✅ | ✅ |
| `trunk` with fixed priority | ✅ | ✅ |

To test "`trunk` with fixed priority", I manually edited `build/blocks/legacy-widget.php` to simulate what will happen when the changes in this PR eventually make its way into Core like so:

```diff
- add_action( 'init', 'register_block_core_legacy_widget', 20 );
+ add_action( 'init', 'register_block_core_legacy_widget' );
```

@kevin940726: Does that make sense? Ready to merge this if so.

**Review by kevin940726 — APPROVED** [review]
Tested it again and not seeing the white screen of death anymore 👍!

**Pull request merged**
